File: Algorithms_Part_2/Ex27FordFulkerson.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.Bag;
import edu.princeton.cs.algs4.FordFulkerson;
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.StdOut;

public class Ex27FordFulkerson {

    private Ex25FlowEdge[] edgeTo;
    private boolean[] marked;
    private double value;

    public Ex27FordFulkerson(Ex26FlowNetwork G, int s, int t) {
        value = 0.0;
        while (hasAugumentingPath(G, s, t)) {
            double bottle = Double.POSITIVE_INFINITY;
            for (int v = t; v != s; v = edgeTo[v].other(v))
                bottle = Math.min(bottle, edgeTo[v].residualCapacityTo(v));

            for (int v = t; v != s; v = edgeTo[v].other(v))
                edgeTo[v].addResidualFlowTo(v, bottle);

            value += bottle;
        }
    }

    public double value() {
        return value;
    }

    public boolean inCut(int v) {
        return marked[v];
    }

    public void printMinCut() {
        Bag<Integer> fromS = new Bag<>();
        Bag<Integer> toT = new Bag<>();
        for (int i = 0; i < marked.length; i++) {
            if (marked[i]) fromS.add(i);
            else toT.add(i);
        }

        StdOut.print("S: ");
        for (int i : fromS)
            StdOut.print(i + " ");
        StdOut.println();
        StdOut.print("T: ");
        for (int i : toT)
            StdOut.print(i + " ");
        StdOut.println();
    }

    private boolean hasAugumentingPath(Ex26FlowNetwork G, int s, int t) {
        edgeTo = new Ex25FlowEdge[G.V()];
        marked = new boolean[G.V()];

        Queue<Integer> queue = new Queue<>();
        queue.enqueue(s);
        marked[s] = true;
        while (!queue.isEmpty()) {
            int v = queue.dequeue();
            for (Ex25FlowEdge e : G.adj(v)) {
                int w = e.other(v);
                if (e.residualCapacityTo(w) > 0 && !marked[w]) {
                    edgeTo[w] = e;
                    queue.enqueue(w);
                    marked[w] = true;
                }
            }
        }

        return marked[t];
    }

    private static void printMinCut(FordFulkerson ff, int V) {
        Bag<Integer> fromS = new Bag<>();
        Bag<Integer> toT = new Bag<>();
        for (int i = 0; i < V; i++) {
            if (ff.inCut(i)) fromS.add(i);
            else toT.add(i);
        }

        StdOut.print("S: ");
        for (int i : fromS)
            StdOut.print(i + " ");
        StdOut.println();
        StdOut.print("T: ");
        for (int i : toT)
            StdOut.print(i + " ");
        StdOut.println();
    }

    public static void main(String[] args) {
        In in = new In("testFN.txt");
        // if (args.length > 0) in = new In(args[0]);
        Ex26FlowNetwork fn = new Ex26FlowNetwork(in);
        StdOut.println(fn);
        Ex27FordFulkerson ff = new Ex27FordFulkerson(fn, 0, fn.V() - 1);
        StdOut.println("value:" + ff.value());
        ff.printMinCut();
        StdOut.println(fn);
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex27FordFulkerson.py
from Ex26FlowNetwork import FlowNetwork


class FordFulkerson:

    def __init__(self, flow_network, s, t):
        self.max_flow = 0.0
        self.edge_to = [None] * flow_network.v
        self.marked = [False] * flow_network.v

        def has_augmenting_path():
            self.edge_to = [None] * flow_network.v
            self.marked = [False] * flow_network.v

            queue = [s]
            self.marked[s] = True
            while len(queue) > 0:
                ve = queue.pop(0)
                self.marked[ve] = True
                for edge in flow_network.get_adj(ve):
                    w = edge.other(ve)
                    if not self.marked[w] and edge.residual_capacity_to(w) > 0:
                        self.edge_to[w] = edge
                        self.marked[w] = True
                        queue.append(w)

            return self.marked[t]

        while has_augmenting_path():
            capacity = float('inf')
            v = t
            while v != s:
                capacity = min(capacity, self.edge_to[v].residual_capacity_to(v))
                v = self.edge_to[v].other(v)

            v = t
            while v != s:
                self.edge_to[v].add_residual_capacity_to(v, capacity)
                v = self.edge_to[v].other(v)

            self.max_flow += capacity

    def in_cut(self, v):
        return self.marked[v]


def test_ford_fulkerson():
    fn = FlowNetwork(0)
    fn.init_from('tinyFN.txt')
    print()
    print(fn)

    ff = FordFulkerson(fn, 0, fn.v - 1)
    print()
    print(f'max_flow: {ff.max_flow}')
    from_s = [num for num in range(fn.v) if ff.in_cut(num)]
    from_t = [num for num in range(fn.v) if not ff.in_cut(num)]
    print(f'S: {from_s}')
    print(f'T: {from_t}')


if __name__ == '__main__':
    test_ford_fulkerson()
